### PR TITLE
Make distance metric configurable in LanceDB

### LINKED ISSUE #41: Expose MetricType in LanceDB
Allow users to customize the metric / distance type when building index and perform queries.

### PR DESCRIPTION
**Pull request by changhiskhan**
Closes #41

### CODE REVIEW

**Comment by changhiskhan** [comment]
i think there's a bug in the cosine distance function in lance

File: python/lancedb/query.py [hunk 1]
```diff
@@ -118,6 +134,7 @@ def to_df(self) -> pd.DataFrame:
                 "column": VECTOR_COLUMN_NAME,
                 "q": self._query,
                 "k": self._limit,
+                "metric": self._metric,
```
Comment (eddyxu) [inline python/lancedb/query.py:137]: does it need to normalize to lower / upper case ?

File: python/lancedb/table.py [hunk 2]
```diff
@@ -121,6 +124,7 @@ def create_index(self, num_partitions=256, num_sub_vectors=96):
         self._dataset.create_index(
             column=VECTOR_COLUMN_NAME,
             index_type="IVF_PQ",
+            metric=metric,
```
Comment (eddyxu) [inline python/lancedb/table.py:127]: do we need to do checks on metric string? or normalize it to upper / lower case?

**Review by eddyxu — APPROVED** [review]
Minor questions.

File: python/lancedb/table.py (hunk 2 above)
Comment (changhiskhan) [reply python/lancedb/table.py:127]: we do this in Lance already. And the Rust code converts it to lowercase

**Pull request merged**